File: algorithms/dfs/dfs.cpp
#include <iostream>
#include <vector>

using namespace std;

void dfs(const vector<vector<int>>& graph,
         int vertex,
         vector<bool>& visited,
         vector<int>& order) {
    visited[vertex] = true;
    order.push_back(vertex);

    for (int neighbor : graph[vertex]) {
        if (!visited[neighbor]) {
            dfs(graph, neighbor, visited, order);
        }
    }
}

int main() {
    vector<vector<int>> graph = {
        {1, 2},
        {0, 3, 4},
        {0, 5},
        {1},
        {1},
        {2}
    };

    vector<bool> visited(graph.size(), false);
    vector<int> order;

    dfs(graph, 0, visited, order);

    cout << "DFS traversal: ";
    for (int vertex : order) cout << vertex << ' ';
    cout << '\n';

    return 0;
}

File: algorithms/dfs/dfs.java
import java.util.*;

public class DFS {
    static void dfs(List<List<Integer>> graph, int vertex,
                    boolean[] visited, List<Integer> order) {
        visited[vertex] = true;
        order.add(vertex);

        for (int neighbor : graph.get(vertex)) {
            if (!visited[neighbor]) {
                dfs(graph, neighbor, visited, order);
            }
        }
    }

    static List<Integer> dfs(List<List<Integer>> graph, int start) {
        boolean[] visited = new boolean[graph.size()];
        List<Integer> order = new ArrayList<>();
        dfs(graph, start, visited, order);
        return order;
    }

    public static void main(String[] args) {
        List<List<Integer>> graph = new ArrayList<>();
        graph.add(Arrays.asList(1, 2));
        graph.add(Arrays.asList(0, 3, 4));
        graph.add(Arrays.asList(0, 5));
        graph.add(Collections.singletonList(1));
        graph.add(Collections.singletonList(1));
        graph.add(Collections.singletonList(2));

        System.out.println("DFS traversal: " + dfs(graph, 0));
    }
}
